_adr/83/index.adoc

```diff
@@ -0,0 +1,131 @@
+. Make sure you have enough quota allocated. This requires:
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (granted/denied) containing all the granted resources or, eventually, any missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API:
++
+```
+/can/{verb}/{resource}
+```
++
+Output of the API will be a JSON with the following structure:
++
+```
+{
+  outcome: string,           // enum: granted/denied
+  metadata: {
+    {key string}:{value}
+  }
+  errors: [{
+    code: string,
+    description: string,
+    metadata: {             // additional details
+      {key string}: {value string}
+    }
+  },...]
+}
+```
++
+Some real world example for an invocation of the `/can/create/kafka` could be:
+
+.. Just one quota available: `standard.x2` won't be returned
++
+```
+{
+  outcome: 'granted',
+  metadata: {
+    plans: ['standard.x1']
+    quota: {
+      granted: 6,
+      used: 5
+    }
+  }
+}
+```
+
+.. More than one quota available:
++
+```
+{
+  outcome: 'granted',
+  metadata: {
+    plans: ['standard.x1', 'standard.x2']
+    quota: {
+      granted: 6,
+      used: 5
+    }
+  }
+}
+```
+
+.. Terms not accepted
++
+```
+{
+  outcome: "denied",
+  errors: [{
+    code: "ERR-TERMS-01",
+    description: "Required terms have not been accepted",
+    metadata: {
+      acceptance_url: "https://url.to.accept.terms"
+    }
+  }]
+}
+```
+
+### Threat model
+
+* No threat model changes are expected here.
```

Comment:
See my message on authentication. Such an API might increase the surface and thus threat.

Reply:
hmmm... The API won't expose anything new: it just aggregates in a single API info's that can already be gathered by querying different subsystems/api.

Reply:
Yeah, I don't think this introduces any meaningful threats. We don't sufficiently rate limit API calls to mean that by calling this API the actor could get around a rate limit, so I think we aren't exposing anything here here that can't be obtained using brute force.